velox/functions/sparksql/tests/FromJsonTest.cpp

```diff
@@ -0,0 +1,250 @@
+      {R"({"a": 1})",
+       R"({"a": 2.0})",
+       R"({"a": "true"})",
+       R"({"a": {"b": "test"}})",
+       R"({"a": [1, 2]})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, nestedComplexType) {
+  auto rowVector = makeRowVector({"a"}, {makeFlatVector<int64_t>({1, 2, 2})});
+  std::vector<vector_size_t> offsets;
+  offsets.push_back(0);
+  offsets.push_back(1);
+  offsets.push_back(2);
+  auto arrayVector = makeArrayVector(offsets, rowVector);
+  auto input = makeFlatVector<std::string>(
+      {R"({"a": 1})", R"([{"a": 2}])", R"([{"a": 2}])"});
+  testFromJson(input, arrayVector);
+}
+
+TEST_F(FromJsonTest, structEmptyArray) {
+  auto expected = makeNullableFlatVector<int64_t>({std::nullopt, 2, 3});
+  auto input =
+      makeFlatVector<std::string>({R"([])", R"({"a": 2})", R"({"a": 3})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, structEmptyStruct) {
+  auto expected = makeNullableFlatVector<int64_t>({std::nullopt, 2, 3});
+  auto input =
+      makeFlatVector<std::string>({R"({ })", R"({"a": 2})", R"({"a": 3})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, structWrongSchema) {
+  auto expected = makeNullableFlatVector<int64_t>({std::nullopt, 2, 3});
+  auto input = makeFlatVector<std::string>(
+      {R"({"b": 2})", R"({"a": 2})", R"({"a": 3})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, structWrongData) {
+  auto expected = makeNullableFlatVector<int64_t>({std::nullopt, 2, 3});
+  auto input = makeFlatVector<std::string>(
+      {R"({"a": 2.1})", R"({"a": 2})", R"({"a": 3})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
+}
+
+TEST_F(FromJsonTest, invalidType) {
+  auto primitiveTypeOutput = makeFlatVector<int64_t>({2, 2, 3});
+  auto dateTypeOutput = makeFlatVector<int32_t>({2, 2, 3}, DATE());
+  auto decimalOutput = makeFlatVector<int64_t>({2, 2, 3}, DECIMAL(16, 7));
+  auto mapOutput =
+      makeMapVector<int64_t, int64_t>({{{1, 1}}, {{2, 2}}, {{3, 3}}});
+  auto input = makeFlatVector<std::string>({R"(2)", R"({2)", R"({3)"});
+  VELOX_ASSERT_USER_THROW(
+      testFromJson(input, primitiveTypeOutput), "Unsupported type BIGINT.");
+  VELOX_ASSERT_USER_THROW(
+      testFromJson(input, makeRowVector({"a"}, {dateTypeOutput})),
+      "Unsupported type ROW<a:DATE>.");
+  VELOX_ASSERT_USER_THROW(
+      testFromJson(input, makeRowVector({"a"}, {decimalOutput})),
+      "Unsupported type ROW<a:DECIMAL(16, 7)>");
+  VELOX_ASSERT_USER_THROW(
+      testFromJson(input, mapOutput), "Unsupported type MAP<BIGINT,BIGINT>.");
+}
+
+TEST_F(FromJsonTest, invalidJson) {
+  auto expected = makeNullableFlatVector<int32_t>(
+      {std::nullopt, std::nullopt, std::nullopt});
+  auto input =
+      makeFlatVector<std::string>({R"("a": 1})", R"({a: 1})", R"({"a" 1})"});
+  testFromJson(input, makeRowVector({"a"}, {expected}));
```

Comment:
Is it the case that if JSON doesn't match the type specified in the "cast", the result is null? Would you document that? 

Test name says "invalidJson" suggesting that it works with malformed JSON, but the JSON in the test seems perfectly valid.

Reply:
The test cases actually contain invalid JSON. Null is returned, because the input string is unparsable. The first case is missing the opening `{`, the second case has a key without double quotes, and the third case is missing a `:`. The valid JSON string is `{"a": 1}`.

Reply:
@zhli1142015 Got it. Thank you for clarifying. What happens if JSON is valid but doesn't match the type? E.g. JSON is an array, but type is a map or a struct?

Reply:
For this case, null is returned, test case `structEmptyArray` covers this.